CONTRIBUTING.md

```diff
@@ -1,36 +1,210 @@
 
-```shell
-git pull upstream master
+```console
+git push
 ```
 
+You are encouraged to push changes frequently (at least once a day if not more often) to your fork.
+This allows other developers to see your code as well as backs up the code on the GitHub servers.
+In addition, pushing code to a fork allows the CI servers to run tests on your code potentially helping find breaking changes sooner.
+
+Feel free to create branches for new features, fixing issues, experiments, etc and have them in your fork.
+Your fork is your workspace to develop freely and experiment.
+Just make sure to use **descriptive and concise** names for your branches so others can identify them.
+
+## Bringing branches up to date
+Changes that happen upstream are not automatically pulled into your fork or your local clone.
+
+One concept to be aware of is that git stores a pool of all the changes (git objects) from the remotes that you are connected to.
+This pool of objects are not kept up to date automatically.
+To fetch all the latest objects from all of your remotes (while also deleting branches that are deleted in remotes) run:
+
+```console
+git remote update -p
+```
+
+### Bringing master up to date
+Based on the recommended workflow the master branch of your local clone should not have any commits that diverge from upstream.
+Because there are not commits that diverge from upstream it is possible to fast forward merge your local master.
+
+To do this first verify you are on master with `git status` or change to master with `git checkout master`.
+
+Then bring your local master up to date by running:
+
+```console
+git merge --ff-only upstream/master
+```
+
+If you get an error that means you have unexpected commits on your master branch that should be removed.
+
+If everything succeeds then you can go ahead and push your updated local master to your fork:
+
+```console
+git push
+```
+
+### Rebasing your branch on top of latest changes
+If you have a branch that was previously created from master you may want to bring the latest changes from master into your branch.
+If you performed a simple `git pull` you will end up with merge commits in your history.
+To avoid the merge commits, the recommended workflow is to instead rebase your local commits in your branch on top of master.
+
+Conceptually, the rebase operation will remove your current commits, bring your branch up to date with master, and then replay your commits one at a time on top of the latest changes.
+You will have the chance to resolve merge conflicts that may occur or safely cancel the rebase operation if desired.
+
+To rebase your current branch on the upstream master branch run:
+```console
+git rebase -p upstream/master
+```
+
+**NOTICE**: Be aware that performing the rebase will cause your local clone of a your branch and the fork version of your branch to diverge.
+In order to push the branch to your clone you have to rewrite the history of the branch in your clone:
```

Comment:
Did you mean "push the branch to your fork"? I thought we're referring to clone in the sentence before to mean the local branch.